In [1]:
import numpy as np
import pandas as pd
import json
from urllib.request import urlopen
import sqlite3


# I- Load the data

### Load Quotebank data

First, let's recover the quotation of interest : as project is based on the caracterisation of the speaker, we decide to pre-select the quotations that are related to a speaker (i.e speaker value is different from 'None'). 

In [4]:
#declaration of a key_world list
key_word = ["carbon dioxide", "greenhouse gas", "global warming",
             "climate change",  "fossil fuels", "sea-level rise",
             "renewable energy", "CO2","methane","PPM","COP","GIEC", 
             "biofuels","business as usual", "carbon footprint", "carbon neutral", "carbon sequestration"] 

In [5]:
def chunk_filtering(chunk, lst):
    template=[] #creation of an empty list :it's always cheaper to append to a list and create a DataFrame than append on a empty dataframe.
    for i in lst: 
        template.append(chunk.loc[chunk["quotation"].apply(lambda x : i in x) & 
                                  chunk["speaker"].apply(lambda x: x!= "None")])#select the quotation with value in speaker column different from 'None' 
                                                                                #and quotations containing the key word
        
    return (pd.concat(template, ignore_index=True))# return a dataframe with our data of interest

##### *2020 quotes extractions*

In [ ]:
i=0
with pd.read_json('data/quotes-2020.json.bz2', lines=True, compression='bz2', chunksize=1000) as df_reader : 
    for chunk in df_reader:
        chunk_clean=chunk_filtering(chunk, key_word)
        header = i == 0
        mode = 'w' if i == 0 else 'a'
        chunk_clean.to_csv(path_or_buf="data/clean_quotes-2020.bz2",compression='bz2',header=header, mode=mode)
        i+=1

In [108]:
quotes_2020= pd.read_csv('data/clean_quotes-2020.bz2', compression='bz2')

In [5]:
quotes_2020.shape

(182024, 10)

-> We extracted 182 024 quotations

##### *2019 quotes extractions*

In [4]:
i=0
with pd.read_json('data/quotes-2019.json.bz2', lines=True, compression='bz2', chunksize=1000) as df_reader:
    for chunk in df_reader:
        header = i == 0
        mode = 'w' if i == 0 else 'a'
        chunk_clean=chunk_filtering(chunk, key_word)
        chunk_clean.to_csv(path_or_buf='data/clean_quotes-2019.bz2', compression='bz2', mode = mode, header=header)
        i+=1
        # create a new csv files compress with bz2 containing all the dataframe recover from the chunk; 

In [5]:
quotes_2019= pd.read_csv('data/clean_quotes-2019.bz2', compression='bz2')
quotes_2019.head(10)
quotes_

Unnamed: 0            quoteID  \
0           0  2019-09-26-057277   
1           1  2019-07-22-037315   
2           0  2019-06-19-067938   
3           0  2019-05-08-019902   
4           0  2019-01-20-042295   

                                           quotation           speaker  \
0  It's managing those zillions of microbial cell...   Thomas Crowther   
1  Malcolm is the right person at the right time ...      Linda Church   
2  Near-zero emission CNG vehicles, paired with r...       Andrew West   
3  Even for Australians who can afford to pay, ge...           Dr Karl   
4  The challenges this year will include global t...  Nguyen Xuan Phuc   

                                                qids                 date  \
0                                      ['Q22958878']  2019-09-26 11:00:45   
1                                       ['Q6551474']  2019-07-22 16:50:00   
2  ['Q19662403', 'Q4758881', 'Q4758888', 'Q570751...  2019-06-19 19:33:00   
3                                       ['Q2063254']  2019-05-08 22:36:00   
4                                      ['Q10800073']  2019-01-20 03:40:26   

   numOccurrences                                             probas  \
0               1  [['Thomas Crowther', '0.9631'], ['None', '0.03...   
1               1   [['Linda Church', '0.6784'], ['None', '0.3216']]   
2               1    [['Andrew West', '0.9237'], ['None', '0.0763']]   
3               5  [['Dr Karl', '0.7559'], ['None', '0.2364'], ['...   
4               7  [['Nguyen Xuan Phuc', '0.8086'], ['None', '0.1...   

                                                urls phase  
0  ['https://www.yaleclimateconnections.org/2019/...     E  
1  ['https://forextv.com/top-news/national-hydrop...     E  
2  ['https://www.naturalgasintel.com/articles/118...     E  
3  ['https://reneweconomy.com.au/climate-change-c...     E  
4  ['https://www.freemalaysiatoday.com/category/b...     E

In [10]:
quotes_2019.shape

(47280, 10)

##### *2018 quotes extractions*

In [9]:
i=0
with pd.read_json('data/quotes-2018.json.bz2', lines=True, compression='bz2', chunksize=1000) as df_reader:
    for chunk in df_reader:
        header = i == 0
        mode = 'w' if i == 0 else 'a'
        chunk_clean=chunk_filtering(chunk, key_word)
        chunk_clean.to_csv(path_or_buf='data/clean_quotes-2018.bz2', compression='bz2', mode = mode, header=header)
        i+=1

EOFError: Compressed file ended before the end-of-stream marker was reached

In [ ]:
quotes_2018= pd.read_csv('data/clean_quotes-2018.bz2', compression='bz2')
quotes_2018.shape

##### *2017 quotes extractions*

In [ ]:
with pd.read_json('data/quotes-2017.json.bz2', lines=True, compression='bz2', chunksize=1000) as df_reader:
    for chunk in df_reader:
        chunk_clean=chunk_filtering(chunk)
        chunk_clean.to_csv(path_or_buf='data/clean_quotes-2017.bz2', compression='bz2', mode = 'a') # create a new csv files compress with bz2 containing all the dataframe recover from the chunk; 

In [ ]:
quotes_2017= pd.read_csv('data/clean_quotes-2018.bz2', compression='bz2')
quotes_2017.shape

##### *2016 quotes extractions*

In [ ]:
with pd.read_json('data/quotes-2016.json.bz2', lines=True, compression='bz2', chunksize=1000) as df_reader:
    for chunk in df_reader:
        chunk_clean=chunk_filtering(chunk)
        chunk_clean.to_csv(path_or_buf='data/clean_quotes-2016.bz2', compression='bz2', mode = 'a') # create a new csv files compress with bz2 containing all the dataframe recover from the chunk; 

In [ ]:
quotes_2016= pd.read_csv('data/clean_quotes-2016.bz2', compression='bz2')
quotes_2016.shape

##### *2015 quotes extractions*

In [4]:
with pd.read_json('data/quotes-2015.json.bz2', lines=True, compression='bz2', chunksize=1000) as df_reader:
    for chunk in df_reader:
        chunk_clean=chunk_filtering(chunk)
        chunk_clean.to_csv(path_or_buf='data/clean_quotes-2015.bz2', compression='bz2', mode = 'a') # create a new csv files compress with bz2 containing all the dataframe recover from the chunk; 

(13829, 10)

In [ ]:
quotes_2015= pd.read_csv('data/clean_quotes-2016.bz2', compression='bz2')
quotes_2015.shape

In [109]:
#additioner tous les quotes

At result we obtained 5 set of Data each with a sufficient size, on total we obtained data to analyze.

## Load additional data Relative to speakers

The provided speaker_attributes.parquet file contains attributes in terms of QIDs, thereby being uninterpretable by humans (df_qid).
To map the QIDs to meaningful labels, we used the provied wikidata_labels_descriptions_quotebank.csv.bz2 containg the labels and value fo the respective QID containing the df_qid (df_label_qid)
By combaning the information of both we can obtained usefule information about speakers. 

In [2]:
df_qid = pd.read_parquet("speaker_attributes.parquet",engine= "pyarrow" )
df_label_qid = pd.read_csv('data/wikidata_labels_descriptions_quotebank.csv.bz2', compression='bz2', index_col='QID')


In [3]:
all(df_qid.academic_degree.notnull())

False

--> Academic_degree value are missing let's remove this column.

In [4]:
df_qid=df_qid.drop(['academic_degree'], axis=1)

In [54]:
df_qid.head()

aliases            date_of_birth  \
0  [Washington, President Washington, G. Washingt...  [+1732-02-22T00:00:00Z]   
1  [Douglas Noel Adams, Douglas Noël Adams, Dougl...  [+1952-03-11T00:00:00Z]   
2      [Paul Marie Ghislain Otlet, Paul Marie Otlet]  [+1868-08-23T00:00:00Z]   
3  [George Walker Bush, Bush Jr., Dubya, GWB, Bus...  [+1946-07-06T00:00:00Z]   
4  [Velázquez, Diego Rodríguez de Silva y Velázqu...  [+1599-06-06T00:00:00Z]   

      nationality      gender   lastrevid ethnic_group US_congress_bio_ID  \
0  [Q161885, Q30]  [Q6581097]  1395141751         None            W000178   
1          [Q145]  [Q6581097]  1395737157   [Q7994501]               None   
2           [Q31]  [Q6581097]  1380367296         None               None   
3           [Q30]  [Q6581097]  1395142029         None               None   
4           [Q29]  [Q6581097]  1391704596         None               None   

                                          occupation      party     id  \
0  [Q82955, Q189290, Q131512, Q1734662, Q294126, ...  [Q327591]    Q23   
1  [Q214917, Q28389, Q6625963, Q4853732, Q1884422...       None    Q42   
2  [Q36180, Q40348, Q182436, Q1265807, Q205375, Q...       None  Q1868   
3  [Q82955, Q15982858, Q18814623, Q1028181, Q1408...   [Q29468]   Q207   
4                                         [Q1028181]       None   Q297   

               label                               candidacy  type  \
0  George Washington                      [Q698073, Q697949]  item   
1      Douglas Adams                                    None  item   
2         Paul Otlet                                    None  item   
3     George W. Bush  [Q327959, Q464075, Q3586276, Q4450587]  item   
4    Diego Velázquez                                    None  item   

                     religion  
0                   [Q682443]  
1                        None  
2                        None  
3  [Q329646, Q682443, Q33203]  
4                        None

In [129]:
df_qid[df_qid.type!='item']

Empty DataFrame
Columns: [aliases, date_of_birth, nationality, gender, lastrevid, ethnic_group, US_congress_bio_ID, occupation, party, academic_degree, id, label, candidacy, type, religion]
Index: []

In [134]:
#Bloc dans lequel j'ai cherché la ligne qui beugue
i=0
for occups in df_qid ['occupation']:
    i=i+1
    if occups is None:
        continue
    for occup in occups: 
        if occup == 'Q3268166': 
            print (i-1)

In [125]:
#Bloc ou je vérifie que cette ligne contient bien le QI manquant (Q3268166)
df_qid ['occupation'][92555]

array(['Q212980', 'Q1650915', 'Q3268166'], dtype=object)

In [137]:
i=0
for occups in df_qid ['candidacy']:
    i=i+1
    if occups is None:
        continue
    for occup in occups: 
        if occup == 'Q20432251': 
            print (i-1)

402813
1535272
2232949
2232977
2656184
2656184
2666879
2667608
2667608
3363864
3787931
4929217
6060780
6063216
7892821
9023827


In [ ]:
#Function that loops through a column to replace QID with their label, and skip None value
def l(df) : 
    liste=[]
    for row  in df: 
        if row is None: 
            continue
        template=[]
        for value in row:
            template.append(df_label_qid.loc[value]['Label'])
        liste.append(template)
    return pd.Series(liste)
#Applying the function to every column containing QIDs (except for occupation, and gender the ones with the faulty QID)
df_qid [['nationality', 'ethnic_group', 'candidacy', 'religion']] = df_qid[['nationality', 'ethnic_group', 'candidacy', 'religion']].apply(l)



# II- Filter the data

As a good data scientist, the first thing to do is to clean up the data . 
move the missing rows; if there are
Let's check if the idenfier is unique, and we haven't duplicate rows;
Let's check if the alias match the label for the df_qid data. 


We also need to extract quotation that refers to our subject of interest : climate change. 
To do so, we decided to creat a list of key_world (based of https://www.climaterealityproject.org/blog/key-terms-you-need-understand-climate-change) and extract contation containing these world. 


##### *verifiy is index is unique*

##### *verifiy is alias mathc with the label*

##### *extraction of interest quotation*

In [8]:
#declaration of a key_world list
key_world = ["carbon dioxide", "greenhouse gas", "global warming",
             "climate change",  "fossil fuels", "sea-level rise",
             "renewable energy", "CO2","methane","PPM","COP","GIEC", 
             "biofuels","business as usual", "carbon footprint", "carbon neutral", "carbon sequestration"] 

In [11]:
#filter_key_world(quotes_2020, key_world).shape
with pd.read_csv('data/clean_quotes-2019.bz2', compression='bz2', chunksize=1000) as df_reader:
    for chunk in df_reader:
        chunk_clean=filter_key_world(chunk, key_world)
        chunk_clean.to_csv(path_or_buf='data/filter_quotes-2019.bz2', compression='bz2', mode = 'a')
        
    

In [12]:
filter_quotes_2019= pd.read_csv('data/filter_quotes-2019.bz2', compression='bz2')
filter_quotes_2019.shape

(56901, 11)

--> le nouveau data contient plus que 427 row (vs 182024 at the begining)

## I-Exploration of our data

Let's see some distribution and statitics: 
 - aged people vs yound people 
 - party politics  
ect... 

stat : correlation coeff ; m